lib/temporal/interceptor/chain.rb

```diff
@@ -0,0 +1,27 @@
+module Temporal
+  module Interceptor
+    class Chain
+      def initialize(interceptors = [])
+        @interceptors = interceptors
+      end
+
+      def invoke(method, input)
+        chain = interceptors.dup
+
+        traverse_chain = lambda do |i|
+          if chain.empty?
+            yield(i)
```

Comment:
@laertispappas you need to have access to both inputs and return values in your interceptor, potentially modifying both and maybe not even not yielding (in case you want to restrict some calls for example). This is why you can't just do something like:

```ruby
chain.each { |interceptor| interceptor.start_workflow(input) }
```

Another potential solution is to inject the reference to the next interceptor when initializing, but that puts restrictions on the shape of the constructor, which I didn't want.

Your example doesn't make much sense because you're yielding twice in the `InterceptorOne`, but the result produced looks correct. Here's a more coherent example:

```ruby
class InterceptorOne
  def start(input)
    input << :before_one
    res = yield(input)
    res << :after_one
    res
  end
end

class InterceptorTwo
  def start(input)
    input << :before_two
    res = yield(input)
    res << :after_two
    res
  end
end

Temporal::Interceptor::Chain.new([InterceptorOne.new, InterceptorTwo.new]).invoke(:start, []) do |input|
  input << :main
  input
end

=> [:before_one, :before_two, :main, :after_two, :after_one]
```

Hope this makes sense

Reply:
@antstorm  Thank you for your feedback much appreciated. One note though to make sure I understand this correctly. From your example above the `after_x` statements will be evaluated only after the main block is executed. Therefore the call stack will most likely result as follows:

```
res = Temporal::Interceptor::Chain.new([One.new, Two.new]).invoke(:start, []) do |input|
  input << :main
   p "inside main result: #{input}"
  input
end
p "after main call result: #{res}"
```

Which will give us:

```
"inside main result: [:before_one, :before_two, :main]"
"after main call result: [:before_one, :before_two, :main, :after_two, :after_one]"
```

Thus the main block above after invoke will only get the results before the `yield` statements and not after the yield ones. I'm trying to understand the expected behaviour here. Is this the intended one? 

Code is very elegant btw 🎖️ awesome work.

Reply:
It is the intended behaviour, yes. Think about typical interceptor implications:

add auth headers (needs to happen before the main handler):
```ruby
class AuthInterceptor
  def start_workflow(input)
    input.headers['AUTH'] = 'some-id'
    yield input
  end
end
```

or latency measurement (happens around the main block)
```ruby
class LogInterceptor
  def start_workflow(input)
    start_time = Time.now
    result = yield input
    Metrics.report('latency.start_workflow', Time.now - start_time)
    result
  end
end
```